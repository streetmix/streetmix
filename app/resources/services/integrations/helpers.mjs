import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
