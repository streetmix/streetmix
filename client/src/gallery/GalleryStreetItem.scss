@use "sass:color";
@import "../../styles/variables";

$selection-highlight-colour: $colour-copper-500;

.gallery-street-item {
  display: inline-block;
  position: relative;
  width: $thumbnail-width;
  height: $thumbnail-height;
  border-radius: $border-radius-medium;

  /* this spacing is so that the selected border is not cut off when calling scrollIntoView() */
  padding: 5px;
  box-sizing: content-box;

  /* reset child box sizing */
  > * {
    box-sizing: border-box;
  }

  &:first-child {
    padding-left: 2px;
  }

  &:last-child {
    padding-right: 2px;
  }
}

.gallery-street-item > a {
  display: block;
  color: black;
  text-decoration: none;
  background: fade-out($sky-colour, 0.2);
  text-align: center;
  white-space: normal;
  cursor: pointer;
  position: relative;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  border-radius: $border-radius-medium;

  canvas {
    position: absolute;
    left: 0;
    top: -7px; // Positioning to cover gaps below ground
    width: $thumbnail-width;
    height: $thumbnail-height;
    transform: scale(1.35); // Crops excess space for most streets
    opacity: 0.45;
    filter: grayscale(0.75);
    transition:
      opacity 45ms,
      filter 45ms;

    // TODO: this causes rendering weirdness in Firefox.
    border-radius: $border-radius-medium;
  }

  &:hover canvas {
    opacity: 1;
    filter: grayscale(0);
  }
}

.gallery-with-owner.gallery-street-item > a canvas {
  top: -17px;
}

.gallery-street-item.gallery-selected > a {
  background: color.adjust($ui-colour, $lightness: -10%);
  box-shadow:
    0 0 0 2px $selection-highlight-colour,
    inset 0 0 0 2px rgb(255 255 255 / 100%);

  .gallery-street-item-canvas > div {
    transform: scale(1.05);
  }

  canvas {
    opacity: 1;
    filter: grayscale(0);
  }
}

.gallery-street-item-canvas {
  position: absolute;
  inset: 2px; // Live within the border area
  border-radius: 2px;
  overflow: hidden;

  > div {
    position: relative;
    width: 100%;
    height: 100%;
    transition: transform 45ms ease-in-out;
    transform-origin: center 80px;

    &:hover {
      transform: scale(1.05);
    }
  }

  .gallery-street-item-ground {
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 22px;
    background-color: #f0f0f0;

    .gallery-with-owner & {
      height: 32px;
    }
  }
}

.gallery-street-item-creator {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 18px;
  text-align: center;
  display: block;
  font-size: 0.8em;
  color: black;
}

.gallery-street-item-date {
  position: absolute;
  left: 20px;
  right: 20px;
  bottom: 5px;
  text-align: center;
  display: block;
  font-size: 0.8em;
  color: black;
  font-weight: 300;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  pointer-events: none;
}

.gallery-street-item-delete {
  position: absolute;
  padding: 0;
  width: 20px;
  height: 20px;
  bottom: 7px;
  right: 7px;
  display: flex !important;
  justify-content: center;
  align-items: center !important;
  line-height: 25px;
  user-select: none;
  background: transparent;
  border: 0;
  border-radius: 50%;
  font-size: 0.65em;

  [dir="rtl"] & {
    left: 7px;
    right: auto;
  }

  &:hover {
    background-color: rgb(255 255 255 / 50%);
  }

  &:active {
    background-color: rgb(255 255 255 / 75%);
  }

  svg {
    color: $colour-turquoise-500 !important;
    width: 12px;
  }

  &:hover svg {
    color: $colour-turquoise-600 !important;
  }
}

.gallery-street-item-error {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  border-radius: $border-radius-medium;
  color: $colour-turquoise-600;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 1em;
}
