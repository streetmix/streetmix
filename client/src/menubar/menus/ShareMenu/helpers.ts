import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
