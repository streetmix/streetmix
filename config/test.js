const port = 8000

module.exports = {
  port: port,
  app_host_port: process.env.APP_DOMAIN || 'localhost:' + port,
  restapi: {
    port: port,
    baseuri: '/api'
  }
}
