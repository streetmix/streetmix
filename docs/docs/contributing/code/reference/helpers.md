---
sidebar_position: 2
---

# Helpers and utilities

:::caution Under construction

This page is a work in progress.

:::

We use the following utilities to solve common programming problems.

## Pseudo-random number generator

If you need a one-time use random number (which will never be stored), there's nothing wrong with JavaScript's built-in `Math.random()`. However, there are times where we need a **seed** for a pseudo-random number generator (PRNG). Seeds generate a random-_looking_ sequence of numbers, but they have the benefit of remaining consistent across renders. For example, this is useful on our sidewalk segment, where we want to generate a random set of pedestrians but we want those pedestrians to remain the same between page refreshes or React component updates.

We can generate seeds by calling `generateRandSeed()`, then by feeding those seeds into the `seedrandom` library:

```js title="random-example.js"
import seedrandom from "seedrandom";
import { generateRandSeed } from "./util/random";

function doThingsWithThePRNG() {
  const seed = generateRandSeed();
  const prng1 = seedrandom(seed);

  // Generate pseudo-random numbers. Each time rng() is called, it returns
  // the next pseudo-random number in the sequence.
  console.log(prng1()); // -> first pseudo-random number
  console.log(prng1()); // -> second pseudo-random number
  // .. etc

  // Create another generator with the same seed
  const prng2 = seedrandom(seed);
  console.log(prng2()); // -> same value as the first call to prng1()
}
```

When `seedrandom()` is called, it returns a function that will return a consistent sequence of numbers each time it is called, for the given `seed`. The sequence returned is local to that function.

:::note

Although the `seed` can be any type of value, like a string, `generateRandSeed()` returns only integer values so that when we store seeds in a data model we can check for type consistency.

:::

In the above example, we call `seedrandom()` twice, with the same `seed`, so that we can create the same sequence of pseudo-random numbers in two different scenarios. This example is contrived because the pseudo-random number generators are used in the same function, and in reality the `seedrandom()` call may happen in separate parts of the application.

**References**

- [Random Seeds, Coded Hints, and Quintillions](http://davidbau.com/archives/2010/01/30/random_seeds_coded_hints_and_quintillions.html) [David Bau]
- [seedrandom.js](https://github.com/davidbau/seedrandom) [GitHub]

## axios

The [Axios](https://github.com/axios/axios) library was introduced as a recommendation over native `fetch` as a way to manage requests to the Streetmix back-end API layer. Additionally, Axios Mock Adapter offers the possibility to mock all API requests for integration testing.

If you work on API-related features, consider refactoring the `fetch` calls to use the API in `.assets/scripts/util/api.js`.
