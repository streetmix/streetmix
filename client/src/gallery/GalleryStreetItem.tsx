import React, { useState, useRef, useLayoutEffect } from 'react'
import { useIntl, FormattedMessage } from 'react-intl'
import { useSelector } from '../store/hooks'
import { useGetUserQuery } from '../store/services/api'
import { getStreetUrl } from '../app/page_url'
import DateTimeRelative from '../app/DateTimeRelative'
import StreetName from '../streets/StreetName'
import { drawStreetThumbnail } from '../streets/thumbnail'
import Icon from '../ui/Icon'
import type { Street } from '@streetmix/types'
import './GalleryStreetItem.scss'

const THUMBNAIL_WIDTH = 180
const THUMBNAIL_HEIGHT = 110
const THUMBNAIL_MULTIPLIER = 0.1 * 2

interface GalleryStreetItemProps {
  street: Street
  showStreetOwner: boolean
  selected: boolean
  allowDelete: boolean
  doSelect: (id: string) => void
  doDelete: (id: string) => void
}

function GalleryStreetItem (
  props: GalleryStreetItemProps
): React.ReactElement | null {
  const {
    street,
    showStreetOwner = true,
    selected = false,
    allowDelete = false,
    doSelect,
    doDelete
  } = props
  const dpi = useSelector((state) => state.system.devicePixelRatio || 1)
  const { data: creatorProfile } = useGetUserQuery(street.creatorId)
  const [isError, setError] = useState<boolean>(false)
  const thumbnailEl = useRef<HTMLCanvasElement>(null)
  const intl = useIntl()

  // Effect hook draws thumbnail to canvas element after mounting
  useLayoutEffect(() => {
    if (thumbnailEl.current === null) return

    const ctx = thumbnailEl.current.getContext('2d')
    if (ctx === null) return

    try {
      drawStreetThumbnail(ctx, street.data.street, {
        // TODO: document magic number 2
        width: THUMBNAIL_WIDTH * 2,
        height: THUMBNAIL_HEIGHT * 2,
        dpi,
        multiplier: THUMBNAIL_MULTIPLIER,
        silhouette: false,
        bottomAligned: false,
        transparentSky: false,
        segmentNamesAndWidths: false,
        streetName: false,
        watermark: false
      })
    } catch (error) {
      setError(true)
    }
  }, [thumbnailEl, street, dpi])

  // Define event handlers
  function handleSelectStreet (event: React.MouseEvent): void {
    event.preventDefault()
    if (event.shiftKey || event.ctrlKey || event.metaKey) return
    doSelect(street.id)
  }

  function handleDeleteStreet (event: React.MouseEvent): void {
    event.preventDefault()
    event.stopPropagation()

    const message = intl.formatMessage(
      {
        id: 'prompt.delete-street',
        defaultMessage:
          'Are you sure you want to permanently delete {streetName}? This cannot be undone.'
      },
      {
        streetName:
          street.name ??
          intl.formatMessage({
            id: 'street.default-name',
            defaultMessage: 'Unnamed St'
          })
      }
    )

    if (window.confirm(message)) {
      doDelete(street.id)
    }
  }

  // Bail if there is no street data
  // TODO: handle data errors better
  if (street.data === undefined) {
    console.error('No street data!', street)
    return null
  }

  const classNames = ['gallery-street-item']
  if (selected) {
    classNames.push('gallery-selected')
  }
  if (showStreetOwner) {
    classNames.push('gallery-with-owner')
  }

  return (
    <div className={classNames.join(' ')}>
      <a href={getStreetUrl(street)} onClick={handleSelectStreet}>
        {/* eslint-disable-next-line multiline-ternary -- Formatting conflicts with prettier */}
        {isError ? (
          <div className="gallery-street-item-error">
            <FormattedMessage
              id="gallery.thumbnail-error"
              defaultMessage="Thumbnail image is not available."
            />
          </div>
        ) : (
          /* TODO: document magic number 2 */
          <div className="gallery-street-item-canvas">
            <div>
              <canvas
                width={THUMBNAIL_WIDTH * dpi * 2}
                height={THUMBNAIL_HEIGHT * dpi * 2}
                ref={thumbnailEl}
              />
              <div className="gallery-street-item-ground" />
            </div>
          </div>
        )}

        <StreetName name={street.name} />

        <span className="gallery-street-item-date">
          <DateTimeRelative value={street.updatedAt} />
        </span>

        {/* Show street creator (owner) or 'Anonymous' */}
        {showStreetOwner && (
          <span className="gallery-street-item-creator">
            {creatorProfile?.displayName ??
              street.creatorId ??
              intl.formatMessage({
                id: 'users.anonymous',
                defaultMessage: 'Anonymous'
              })}
          </span>
        )}
      </a>

      {/* Only show delete button if allowed, e.g. if user is owner of the street */}
      {allowDelete && (
        <button
          className="gallery-street-item-delete"
          onClick={handleDeleteStreet}
          title={intl.formatMessage({
            id: 'gallery.delete-street-tooltip',
            defaultMessage: 'Delete street'
          })}
        >
          <Icon icon="trash" />
        </button>
      )}
    </div>
  )
}

export default GalleryStreetItem
