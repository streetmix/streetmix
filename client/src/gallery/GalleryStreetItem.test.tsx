import React from 'react'
import { vi } from 'vitest'
import { screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'

import { render } from '~/test/helpers/render'
import { MOCK_STREET } from '~/test/fixtures'
import GalleryStreetItem from './GalleryStreetItem'

// Mock dependencies
vi.mock('../streets/thumbnail', () => ({
  drawStreetThumbnail: vi.fn()
}))
vi.mock('../app/page_url', () => ({
  getStreetUrl: vi.fn()
}))

const baseProps = {
  street: MOCK_STREET,
  showStreetOwner: true,
  selected: false,
  allowDelete: true,
  doSelect: () => {},
  doDelete: () => {}
}

describe('GalleryStreetItem', () => {
  it('renders', () => {
    // This uses jsdom + canvas packages under the hood to render
    const { asFragment } = render(<GalleryStreetItem {...baseProps} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('does not display street owner when we ask it not to', () => {
    render(<GalleryStreetItem {...baseProps} showStreetOwner={false} />)

    // This is a hard-coded value, it will not be null in this test
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(screen.queryByText(MOCK_STREET.creatorId!)).not.toBeInTheDocument()
  })

  it('displays "Unnamed St" without a street name', () => {
    render(
      <GalleryStreetItem
        {...baseProps}
        street={{
          ...MOCK_STREET,
          name: null
        }}
      />
    )

    expect(screen.getByText('Unnamed St')).toBeInTheDocument()
  })

  it('displays "Anonymous" for anonymous streets', async () => {
    render(
      <GalleryStreetItem
        {...baseProps}
        street={{
          ...MOCK_STREET,
          creatorId: null
        }}
      />
    )

    await waitFor(() => {
      expect(screen.getByText('Anonymous')).toBeInTheDocument()
    })
  })

  it('handles select', async () => {
    const doSelect = vi.fn()
    render(<GalleryStreetItem {...baseProps} doSelect={doSelect} />)

    // This is a hard-coded value, it will not be null in this test
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    await userEvent.click(screen.getByText(MOCK_STREET.name!))
    expect(doSelect).toBeCalled()
  })

  it('handles delete when confirmed', async () => {
    const doDelete = vi.fn()
    window.confirm = vi.fn(() => true)

    render(<GalleryStreetItem {...baseProps} doDelete={doDelete} />)

    await userEvent.click(screen.getByTitle('Delete street'))
    expect(doDelete).toBeCalled()
  })

  it('does not delete when confirmation is cancelled', async () => {
    const doDelete = vi.fn()
    window.confirm = vi.fn(() => false)

    render(<GalleryStreetItem {...baseProps} doDelete={doDelete} />)

    await userEvent.click(screen.getByTitle('Delete street'))
    expect(doDelete).not.toBeCalled()
  })
})
