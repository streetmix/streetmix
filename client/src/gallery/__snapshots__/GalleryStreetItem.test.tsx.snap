// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GalleryStreetItem > renders 1`] = `
<DocumentFragment>
  <div
    class="gallery-street-item"
  >
    <a>
      <div
        class="gallery-street-item-inner"
      >
        <div
          class="gallery-street-item-canvas"
        >
          <canvas
            height="220"
            width="360"
          />
        </div>
        <div
          class="gallery-street-item-label"
        >
          <div>
            foo
          </div>
          <div
            class="gallery-street-item-date"
          >
            <time
              datetime="2018-04-27T20:47:03.477Z"
              title="2018-04-27T20:47:03.477Z"
            >
              April 27, 2018
            </time>
          </div>
        </div>
      </div>
      <div
        class="street-name"
      >
        <div
          class="street-name-text"
        >
          Baz
        </div>
      </div>
    </a>
    <button
      class="gallery-street-item-delete"
      title="Delete street"
    >
      <svg
        data-icon="trash"
        data-icon-source="test"
      />
    </button>
  </div>
</DocumentFragment>
`;
